{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py:3071: DtypeWarning: Columns (4,15,40,41,43,44,74,83,84,85,86,87,97,147) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import json\n",
    "from jordan_plus_genes.model.Sample_MIL import InstanceModels, RaggedModels\n",
    "from jordan_plus_genes.model.KerasLayers import Losses, Metrics\n",
    "import tensorflow as tf\n",
    "from sklearn.metrics import roc_auc_score, classification_report, confusion_matrix\n",
    "from sklearn.model_selection import StratifiedShuffleSplit, StratifiedKFold\n",
    "from jordan.model import DatasetsUtils\n",
    "import pickle\n",
    "import logomaker\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import fisher_exact\n",
    "\n",
    "#physical_devices = tf.config.experimental.list_physical_devices('GPU')\n",
    "#tf.config.experimental.set_memory_growth(physical_devices[-4], True)\n",
    "#tf.config.experimental.set_visible_devices(physical_devices[-4], 'GPU')\n",
    "\n",
    "# Load the raw data files\n",
    "D, tcga_maf, samples = pickle.load(open('/home/janaya2/Desktop/ATGC_paper/figures/tumor_classification/data/data.pkl', 'rb'))\n",
    "\n",
    "cancerhotspots_df = pd.read_csv(\"/home/sahn33/Documents/cancerhotspots.v2.maf\",sep=\"\\t\", low_memory=True) #usecols=[\"Chromosome\",\"Start_Position\", \"End_Position\", \"Reference_Allele\",\"Tumor_Seq_Allele2\"],\n",
    "\n",
    "with open(\"publication_hotspots.vcf\", \"r\") as f:\n",
    "    lines = f.readlines()\n",
    "    chrom_index = [i for i, line in enumerate(lines) if line.strip().startswith(\"#CHROM\")]\n",
    "    data = lines[chrom_index[0]:]\n",
    "header = data[0].strip().split(\"\\t\")\n",
    "informations = [d.strip().split(\"\\t\") for d in data[1:]]\n",
    "publication_hotspots_df = pd.DataFrame(informations, columns=header)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Create a dataframe that identifies the instances that are hotspots.\n",
    "cancerhotspots_df.drop_duplicates(subset=[\"Chromosome\",\"Start_Position\", \"End_Position\", \"Reference_Allele\",\"Tumor_Seq_Allele2\"],keep=False,inplace=True)\n",
    "\n",
    "cancerhotspots_df = cancerhotspots_df[[\"Chromosome\",\"Start_Position\", \"End_Position\", \"Reference_Allele\",\"Tumor_Seq_Allele2\"]] #\n",
    "cancerhotspots_df[\"id\"] = cancerhotspots_df.index\n",
    "\n",
    "hotspot_df_ = pd.merge(cancerhotspots_df, tcga_maf, how='right', on=[\"Chromosome\",\"Start_Position\", \"End_Position\", \"Reference_Allele\",\"Tumor_Seq_Allele2\"], suffixes=('_duplicate',''))\n",
    "\n",
    "publication_hotspots_df.drop_duplicates(subset=[\"#CHROM\",\"POS\",\"REF\",\"ALT\"],keep=False,inplace=True)\n",
    "\n",
    "hotspot_df_copy = hotspot_df_\n",
    "hotspot_df_copy = hotspot_df_copy[[\"Chromosome\",\"Start_Position\",\"Ref\",\"Alt\",\"id\"]] #\n",
    "hotspot_df_copy['Alt'] = hotspot_df_copy['Alt'].apply(lambda x: x[:1])\n",
    "hotspot_df_copy['Ref'] = hotspot_df_copy['Ref'].apply(lambda x: x[:1])\n",
    "hotspot_df_copy['Start_Position'] = hotspot_df_copy['Start_Position'].apply(lambda x: str(x))\n",
    "publication_hotspots_df[\"my_index\"] = publication_hotspots_df.index\n",
    "hotspot_df_merge_cols = [\"Chromosome\",\"Start_Position\",\"Ref\",\"Alt\"]\n",
    "publication_hotspots_df_merge_cols = [\"#CHROM\",\"POS\",\"REF\",\"ALT\"]\n",
    "\n",
    "hotspot_df = pd.merge(publication_hotspots_df, hotspot_df_copy, how='right', right_on = hotspot_df_merge_cols, left_on = publication_hotspots_df_merge_cols, suffixes=('_duplicate',''))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a column in the dataframe that flags for InDel instances\n",
    "\n",
    "tcga_maf['indel'] = np.where((tcga_maf['Variant_Classification'] == 'Frame_Shift_Del') |\n",
    "                             (tcga_maf['Variant_Classification'] == 'Frame_Shift_Ins') |\n",
    "                             (tcga_maf['Variant_Classification'] == 'In_Frame_Del') |\n",
    "                             (tcga_maf['Variant_Classification'] == 'In_Frame_Ins'),\n",
    "                             True, False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.8/dist-packages/pandas/core/indexing.py:1719: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_column(loc, value, pi)\n"
     ]
    }
   ],
   "source": [
    "# Re-assign labels from TCGA to custom NCIT labels and encode the data\n",
    "\n",
    "tcga_maf['Hugo_Symbol'] = tcga_maf['Hugo_Symbol'].astype('category')\n",
    "samples['type'] = samples['type'].apply(lambda x: 'COAD' if x == 'READ' else x)\n",
    "\n",
    "###\n",
    "# filtering the NCI-T labels (https://livejohnshopkins-my.sharepoint.com/:x:/r/personal/abaras1_jh_edu/_layouts/15/doc2.aspx?sourcedoc=%7B5f92f0fc-ec6c-40d5-ab17-0d3345f9f2c2%7D&action=edit&activeCell=%27Sheet1%27!B21&wdinitialsession=e072a38f-57c8-4c1f-885b-efaefcc81d35&wdrldsc=2&wdrldc=1&wdrldr=AccessTokenExpiredWarning%2CRefreshingExpiredAccessT)\n",
    "ncit_labels_kept = ['Muscle-Invasive Bladder Carcinoma','Infiltrating Ductal Breast Carcinoma',\n",
    "                    'Invasive Lobular Breast Carcinoma','Cervical Squamous Cell Carcinoma',\n",
    "                    'Colorectal Adenocarcinoma','Glioblastoma','Head and Neck Squamous Cell Carcinoma',\n",
    "                    'Clear Cell Renal Cell Carcinoma','Papillary Renal Cell Carcinoma','Astrocytoma',\n",
    "                    'Oligoastrocytoma','Oligodendroglioma','Hepatocellular Carcinoma','Lung Adenocarcinoma',\n",
    "                    'Lung Squamous Cell Carcinoma','Ovarian Serous Adenocarcinoma','Adenocarcinoma, Pancreas',\n",
    "                    'Paraganglioma','Pheochromocytoma','Prostate Acinar Adenocarcinoma','Colorectal Adenocarcinoma',\n",
    "                    'Desmoid-Type Fibromatosis','Leiomyosarcoma','Liposarcoma','Malignant Peripheral Nerve Sheath Tumor',\n",
    "                    'Myxofibrosarcoma','Synovial Sarcoma','Undifferentiated Pleomorphic Sarcoma',\n",
    "                    'Cutaneous Melanoma','Gastric Adenocarcinoma','Testicular Non-Seminomatous Germ Cell Tumor',\n",
    "                    'Testicular Seminoma','Thyroid Gland Follicular Carcinoma','Thyroid Gland Papillary Carcinoma',\n",
    "                    'Endometrial Endometrioid Adenocarcinoma','Endometrial Serous Adenocarcinoma']\n",
    "idx_filter = samples['NCI-T Label'].isin(ncit_labels_kept)\n",
    "ncit_samples = samples.loc[idx_filter]\n",
    "PCPG_ncit = ['Paraganglioma','Pheochromocytoma']\n",
    "SARC_ncit = ['Desmoid-Type Fibromatosis','Leiomyosarcoma','Liposarcoma','Malignant Peripheral Nerve Sheath Tumor',\n",
    "             'Myxofibrosarcoma','Synovial Sarcoma','Undifferentiated Pleomorphic Sarcoma']\n",
    "TGCT_ncit = ['Testicular Non-Seminomatous Germ Cell Tumor','Testicular Seminoma']\n",
    "ncit_samples.loc[ncit_samples['NCI-T Label'].isin(PCPG_ncit), 'NCI-T Label'] = 'PCPG'\n",
    "ncit_samples.loc[ncit_samples['NCI-T Label'].isin(SARC_ncit), 'NCI-T Label'] = 'SARC'\n",
    "ncit_samples.loc[ncit_samples['NCI-T Label'].isin(TGCT_ncit), 'NCI-T Label'] = 'TGCT'\n",
    "\n",
    "#samples = ncit_samples\n",
    "A = ncit_samples['NCI-T Label'].astype('category')\n",
    "###\n",
    "\n",
    "D['genes'] = np.concatenate(tcga_maf[['Hugo_Symbol']].apply(lambda x: x.cat.codes).values + 1)\n",
    "input_dim = max(D['genes'])\n",
    "dropout = .5\n",
    "\n",
    "indexes = [np.where(D['sample_idx'] == idx) for idx in np.arange(samples.shape[0])[idx_filter]]\n",
    "#indexes = [np.where(D['sample_idx'] == idx) for idx in np.arange(samples.shape[0])]  #<-- uncomment for non NCIT labels\n",
    "genes = np.array([D['genes'][i] for i in indexes], dtype='object')\n",
    "hotspots = np.array([~pd.isna(hotspot_df[\"my_index\"]).values[i] for i in indexes], dtype='object')\n",
    "indels = np.array([tcga_maf['indel'].values[i] for i in indexes], dtype='object')\n",
    "\n",
    "index_loader = DatasetsUtils.Map.FromNumpytoIndices([j for i in indexes for j in i], dropout=dropout)\n",
    "genes_loader = DatasetsUtils.Map.FromNumpyandIndices(genes, tf.int16)\n",
    "genes_loader_eval = DatasetsUtils.Map.FromNumpy(genes, tf.int16, dropout=0)\n",
    "\n",
    "\n",
    "#A = samples['type'].astype('category') #<-- uncomment for non NCIT labels\n",
    "classes = A.cat.categories.values\n",
    "##integer values for random forest\n",
    "classes_onehot = np.eye(len(classes))[A.cat.codes]\n",
    "y_label = classes_onehot\n",
    "\n",
    "y_strat = np.argmax(y_label, axis=-1)\n",
    "class_counts = dict(zip(*np.unique(y_strat, return_counts=True)))\n",
    "y_weights = np.array([1 / class_counts[_] for _ in y_strat])\n",
    "y_weights /= np.sum(y_weights)\n",
    "\n",
    "y_label_loader = DatasetsUtils.Map.FromNumpy(y_label, tf.float32)\n",
    "y_weights_loader = DatasetsUtils.Map.FromNumpy(y_weights, tf.float32)\n",
    "\n",
    "predictions = []\n",
    "test_idx = []\n",
    "weights = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the models' results.\n",
    "\n",
    "test_idx, weights = pickle.load(open('/home/mlee276/Desktop/TCGA-ML-main/results/mil_gene_weights.pkl', 'rb'))\n",
    "\n",
    "gene_encoder = InstanceModels.GeneEmbed(shape=(), input_dim=input_dim, dim=256)\n",
    "mil = RaggedModels.MIL(instance_encoders=[gene_encoder.model], sample_encoders=[], output_dims=[y_label.shape[-1]], output_types=['other'], mil_hidden=[256], attention_layers=[64, 16], dropout=.5, instance_dropout=.5, regularization=0, input_dropout=dropout)\n",
    " \n",
    "attention_weights = []\n",
    "embedding_weights = []\n",
    "idx_test_all = []\n",
    "for idx_test, fold_weights in zip(test_idx, weights):\n",
    "    mil.model.set_weights(fold_weights)\n",
    "    ds_test = tf.data.Dataset.from_tensor_slices(((\n",
    "                                                       genes_loader_eval(idx_test),\n",
    "                                                   ),\n",
    "    ))\n",
    "    ds_test = ds_test.batch(len(idx_test), drop_remainder=False)\n",
    "    attention_weights.extend(mil.attention_model.predict(ds_test).to_list())\n",
    "    idx_test_all.extend(idx_test)\n",
    "    embedding_weights.append(mil.model.get_weights())\n",
    "\n",
    " ### can only call this once or else the order gets mixed. \n",
    "hotspots_reordered = hotspots[idx_test_all] \n",
    "indels_reordered = indels[idx_test_all]\n",
    "genes_reordered = genes[idx_test_all]   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Get MIL predictions\n",
    "test_idx, mil_predictions = pickle.load(open('/home/mlee276/Desktop/TCGA-ML-main/results/mil_gene_predictions.pkl', 'rb'))\n",
    "mil_predictions_labels = np.argmax(mil_predictions, axis=-1)\n",
    "# Get true labels\n",
    "y_strat = np.argmax(y_label, axis=-1)\n",
    "correct = (y_strat[np.concatenate(test_idx)])\n",
    "\n",
    "### can only call this once or else the order gets mixed. \n",
    "#hotspots_reordered = hotspots[idx_test_all] \n",
    "#indels_reordered = indels[idx_test_all]\n",
    "#genes_reordered = genes[idx_test_all]\n",
    "\n",
    "# Returns the attention weights, hotspot attention weights, and non-hotspot attention weights associated with a specifc cancer type and context sequence.\n",
    "def get_position_weight_hotspots(class_num, sequence):\n",
    "    # Get indexes of correct predictions\n",
    "    correct_predictions_idx = []\n",
    "    for i in range(len(correct)):\n",
    "        if correct[i] == mil_predictions_labels[i] and correct[i] == class_num:\n",
    "            correct_predictions_idx.append(i)\n",
    "            \n",
    "    all_sample_attention_weights = []\n",
    "    hotspot_attention_weights = []\n",
    "    non_hotspot_attention_weights = []\n",
    "    print(len(correct_predictions_idx))\n",
    "    for idx in correct_predictions_idx:\n",
    "        sample_attention_weights = np.concatenate(attention_weights[idx])\n",
    "        \n",
    "        hotspot_attention_weights.append(np.expand_dims(sample_attention_weights, axis=0)[:,hotspots_reordered[idx]][0])\n",
    "        non_hotspot_attention_weights.append(np.expand_dims(sample_attention_weights, axis=0)[:,~hotspots_reordered[idx]][0])\n",
    "        \n",
    "        all_sample_attention_weights.append(sample_attention_weights)\n",
    "    \n",
    "    return all_sample_attention_weights, hotspot_attention_weights, non_hotspot_attention_weights\n",
    "\n",
    "# Returns the attention weights, indel attention weights, and non-indel attention weights associated with a specifc cancer type and context sequence.\n",
    "def get_position_weight_indels(class_num, sequence):\n",
    "    # Get indexes of correct predictions\n",
    "    correct_predictions_idx = []\n",
    "    for i in range(len(correct)):\n",
    "        if correct[i] == mil_predictions_labels[i] and correct[i] == class_num:\n",
    "            correct_predictions_idx.append(i)\n",
    "            \n",
    "    all_sample_attention_weights = []\n",
    "    indel_attention_weights = []\n",
    "    non_indel_attention_weights = []\n",
    "    for idx in correct_predictions_idx:\n",
    "        sample_attention_weights = np.concatenate(attention_weights[idx])\n",
    "        \n",
    "        indel_attention_weights.append(np.expand_dims(sample_attention_weights, axis=0)[:,indels_reordered[idx]][0])\n",
    "        non_indel_attention_weights.append(np.expand_dims(sample_attention_weights, axis=0)[:,~indels_reordered[idx]][0])\n",
    "        \n",
    "        all_sample_attention_weights.append(sample_attention_weights)\n",
    "    \n",
    "    return all_sample_attention_weights, indel_attention_weights, non_indel_attention_weights\n",
    "\n",
    "'''\n",
    "attention_weights: [8920 samples, ragged length genes, 1 float value attention weight]\n",
    "'''\n",
    "# Returns the attention weight and associated instance gene of a specific cancer class where the attention weight is between the percent_lower and percent_upper bounds.\n",
    "# This function can run on sample or instance level data, indicated by the use_sample_level boolean parameter.\n",
    "def get_weights_and_genes(class_num, percent_lower, percent_upper, use_sample_level):\n",
    "    # Get indexes of correct predictions of the specific cancer type\n",
    "    correct_predictions_idx = []\n",
    "    for i in range(len(correct)):\n",
    "        if correct[i] == mil_predictions_labels[i] and correct[i] == class_num:\n",
    "            correct_predictions_idx.append(i)\n",
    "    \n",
    "    upper_thresh = None\n",
    "    lower_thresh = None\n",
    "    class_attention_weights = []\n",
    "    for idx in correct_predictions_idx:\n",
    "        class_attention_weights.append(attention_weights[idx])\n",
    "    all_weights = np.concatenate(class_attention_weights, axis=0).T[0]\n",
    "    all_weights_sorted = np.sort(np.array(all_weights))\n",
    "    # Calculate threshold for attention weights (in respects to the cancer type level)\n",
    "    if not use_sample_level:\n",
    "        upper_thresh = all_weights_sorted[round(len(all_weights)*percent_upper)-1]\n",
    "        lower_thresh = all_weights_sorted[round(len(all_weights)*percent_lower)]\n",
    "    \n",
    "    # For all the correctly predicted samples of this cancer, get the genes with attention weights above the threshold.\n",
    "    filtered_sample_attention_weight_genes = []\n",
    "    for idx in correct_predictions_idx:\n",
    "        # for the correctly predicted sample, get all its attention weights.\n",
    "        sample_attention_weights = np.concatenate(attention_weights[idx])\n",
    "        # Calculate threshold for attention weights (in respects to the sample level)\n",
    "        if use_sample_level:\n",
    "            sample_attention_weights_sorted = np.sort(sample_attention_weights)\n",
    "            upper_thresh = sample_attention_weights_sorted[round(len(sample_attention_weights)*percent_upper)-1]\n",
    "            lower_thresh = sample_attention_weights_sorted[round(len(sample_attention_weights)*percent_lower)]\n",
    "        # get the genes associated with each sample which have above threshold attention weight.\n",
    "        attention_filter = []\n",
    "        for i in range(len(sample_attention_weights)):\n",
    "            if sample_attention_weights[i] >= lower_thresh and sample_attention_weights[i] <= upper_thresh:\n",
    "                attention_filter.append(True)\n",
    "            else:\n",
    "                attention_filter.append(False)\n",
    "        #getting the targeted genes.\n",
    "        filtered_sample_attention_weight_genes.append(genes_reordered[idx][attention_filter])\n",
    "\n",
    "    # filtered_sample_attention_weight_genes: [correctly predicted samples, ragged genes above attention weight threshold]\n",
    "    return all_weights, filtered_sample_attention_weight_genes\n",
    "\n",
    "# Returns the genes and associated frequency of the genes of the gene of a specific cancer class that the model used. The genes are ordered by frequency. \n",
    "# The genes are pre-selected from the entire set of samples for the specific cancer class by the get_weights_and_genes function. \n",
    "# This function can run on sample or instance level data, indicated by the use_sample_level boolean parameter.\n",
    "# This function can be used to return only the top 10 most frequent genes according to the frequency_filter boolean parameter. \n",
    "def get_top_gene_labels(class_num, percent_lower, percent_upper, frequency_filter, use_sample_level):\n",
    "    top_gene_labels = []\n",
    "    # Get indexes of top genes in this cancer class by attention weight\n",
    "    _,filtered_sample_attention_weight_genes = get_weights_and_genes(class_num, percent_lower, percent_upper, use_sample_level)\n",
    "    # Get count of found genes\n",
    "    temp = []\n",
    "    for l in filtered_sample_attention_weight_genes:\n",
    "        temp.extend(l)\n",
    "    filtered_genes = np.array(temp)\n",
    "    \n",
    "    unique_genes, counts = np.unique(filtered_genes, return_counts = True)\n",
    "    counts_sorted_idx = np.argsort(counts)\n",
    "    sorted_filtered_genes = unique_genes[counts_sorted_idx]\n",
    "    top_genes_counted = sorted_filtered_genes\n",
    "    top_genes_counts = counts[counts_sorted_idx]\n",
    "    if frequency_filter:\n",
    "        top_genes_counted = sorted_filtered_genes[-10:]\n",
    "        top_genes_counts = counts[counts_sorted_idx][-10:]\n",
    "    #shift codes by one to get the correct gene label from the code:\n",
    "    top_genes_counted = [n-1 for n in top_genes_counted]\n",
    "    \n",
    "    # make \"selected\" gene idx to label dictionary\n",
    "    top_gene_labels = tcga_maf['Hugo_Symbol'].cat.categories[top_genes_counted]\n",
    "    #cat codes are index categories\n",
    "    \n",
    "    # get cancer type label\n",
    "    cancer_label = classes[class_num]\n",
    "    \n",
    "    return top_gene_labels, cancer_label, top_genes_counts\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'PIK3R1': 0.0, 'ARID1A': 0.0, 'IDH2': 0.0, 'PIK3CA': 0.0, 'NOTCH1': 0.0, 'FUBP1': 0.0, 'IDH1': 0.0, 'CIC': 0.0, 'ATRX': 0.0001, 'NRAS': 0.00099, 'TP53': 0.00125, 'PBRM1': 0.00997, 'CTNNB1': 0.00997, 'ZBTB20': 0.08924, 'ACTC1': 0.1, 'KY': 0.1, 'VN1R4': 0.1, 'LCP1': 0.1, 'MUC4': 0.1, 'OR10S1': 0.1, 'KEL': 0.1, 'OTOL1': 0.1, 'PTEN': 0.1, 'RASGRP3': 0.1, 'RNF43': 0.1, 'SIK2': 0.1, 'SLC6A20': 0.1, 'TCF7L2': 0.1, 'UBR5': 0.1, 'PCDH11Y': 0.1, 'ACTRT1': 0.1, 'GHR': 0.1, 'FOXA1': 0.1, 'FBXO42': 0.1, 'EP300': 0.1, 'EFEMP1': 0.1, 'DNAH6': 0.1, 'ATP1B1': 0.1, 'CAST': 0.1, 'CNOT4': 0.1, 'CDK8': 0.1, 'MTOR': 0.19003, 'MYO5A': 0.19003, 'RAD21': 0.19003, 'IKZF3': 0.19003, 'LAMP2': 0.19003, 'ARID1B': 0.34403, 'BCOR': 0.34403, 'NIPBL': 1.0}\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "we want to compare the count of genes in the top 10% of attention weights and the lower \n",
    "90% of attention weights using a Fisher Exact test to see if there is a relationship.\n",
    "Note: no top 10 frequency of genes filter.\n",
    "'''\n",
    "# Returns the Exact Fisher Score of genes from the get_top_gene_labels function. \n",
    "# The two catagories being compared in the Fisher test are created using the min_perc_1, max_perc_1, min_perc_2, max_perc_2 parameters. The two catagories separate the genes into two sets of genes based on their attention weight. \n",
    "def fisher(cancer_type, min_perc_1, max_perc_1, min_perc_2, max_perc_2):\n",
    "    gene_labels_10, cancer_label_10, gene_counts_10 = get_top_gene_labels(cancer_type, min_perc_1, max_perc_1, frequency_filter=False, use_sample_level=False)\n",
    "    gene_labels_90, cancer_label_90, gene_counts_90 = get_top_gene_labels(cancer_type, min_perc_2, max_perc_2, frequency_filter=False, use_sample_level=False)\n",
    "\n",
    "    gene_labels_counts_10 = dict(zip(gene_labels_10, gene_counts_10))\n",
    "    gene_labels_counts_90 = dict(zip(gene_labels_90, gene_counts_90))\n",
    "\n",
    "    gene_counts_10_sum = np.sum(np.array(gene_counts_10))\n",
    "    gene_counts_90_sum = np.sum(np.array(gene_counts_90))\n",
    "\n",
    "    genes_unioned = np.union1d(np.array(gene_labels_10), np.array(gene_labels_90))\n",
    "    gene_fisher = {}\n",
    "    gene_m = {}\n",
    "    for gene in genes_unioned:\n",
    "        m = [[0,0],[0,0]]\n",
    "        if gene in gene_labels_counts_10:\n",
    "            m[0][0] = gene_labels_counts_10[gene]\n",
    "            m[0][1] = gene_counts_10_sum - gene_labels_counts_10[gene]\n",
    "        else:\n",
    "            m[0][0] = 0\n",
    "            m[0][1] = gene_counts_10_sum\n",
    "        if gene in gene_labels_counts_90:\n",
    "            m[1][0] = gene_labels_counts_90[gene]\n",
    "            m[1][1] = gene_counts_90_sum - gene_labels_counts_90[gene]\n",
    "        else:\n",
    "            m[1][0] = 0\n",
    "            m[1][1] = gene_counts_90_sum\n",
    "        oddsr, p = fisher_exact(m)\n",
    "        gene_fisher[gene] = round(p, 5)\n",
    "        gene_m[gene] = m\n",
    "\n",
    "    gene_10_fisher = {}\n",
    "    for gene in gene_labels_10:\n",
    "        if gene in gene_fisher:\n",
    "            gene_10_fisher[gene] = gene_fisher[gene]\n",
    "\n",
    "    gene_10_fisher = dict(sorted(gene_10_fisher.items(), key=lambda item: item[1]))\n",
    "\n",
    "    return gene_fisher, gene_10_fisher, gene_m\n",
    "    \n",
    "'''gene_fisher, gene_10_fisher, gene_m = fisher(17, 0.9, 1.0, 0.0, 0.9)\n",
    "plt.hist(gene_10_fisher.values(),bins=20)\n",
    "plt.xlabel(\"fisher score\")\n",
    "plt.ylabel(\"freq\")\n",
    "\n",
    "print(gene_10_fisher)'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The top 10 of the top 10% genes based on attention weight for each cancer type:\n",
      "Using matplotlib backend: Qt5Agg\n"
     ]
    }
   ],
   "source": [
    "# Produces a plot of the genes of a cancer type that are in the top x percent of attention weights and have a gene count greater than y.\n",
    "print(\"The top 10 of the top 10% genes based on attention weight for each cancer type:\")\n",
    "table_data = []\n",
    "for i in range(len(classes)):\n",
    "    top_gene_labels, cancer_label, gene_counts = get_top_gene_labels(i, 0.9, 1.0, True, False)\n",
    "    gene_fisher, gene_10_fisher, gene_m = fisher(i, 0.9, 1.0, 0.0, 0.9)\n",
    "    \n",
    "    top_gene_label =  [str(list(top_gene_labels)[j]) + \n",
    "                        \" (\" +  str(list(gene_counts)[j]) + \")\\n\" +\n",
    "                        str('\\n'.join(pd.DataFrame(np.array(gene_m[top_gene_labels[j]])).to_string(index = False).split('\\n')[1:])) +\n",
    "                       \"\\np:\" + str(gene_fisher[top_gene_labels[j]])\n",
    "                        for j in range(len(list(top_gene_labels)))]\n",
    "    row = [cancer_label]\n",
    "    row.extend(top_gene_label)\n",
    "    while len(row) < 11:\n",
    "        row.append(\"\")\n",
    "    table_data.append(np.array(row))\n",
    "    \n",
    "%matplotlib\n",
    "plt.figure(figsize=(16,8))\n",
    "table = plt.table(cellText=table_data,loc=\"center\") #, colWidths=widths)\n",
    "table.auto_set_font_size(False)\n",
    "table.set_fontsize(5)\n",
    "table.scale(1,2)\n",
    "plt.axis(\"off\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using matplotlib backend: Qt5Agg\n"
     ]
    }
   ],
   "source": [
    "# Produces a Matrix of histogram of the Fisher score distribution of the top 10% attention weighted genes for each cancer type\n",
    "%matplotlib\n",
    "fig,axs = plt.subplots(5,6)\n",
    "for i in range(len(classes)):\n",
    "    gene_fisher, gene_10_fisher, gene_m = fisher(i, 0.9, 1.0, 0.0, 0.9)\n",
    "    plt.hist(gene_10_fisher.values(),bins=20)\n",
    "\n",
    "    bin_nums = 20\n",
    "    #plotting\n",
    "    axs[i//6,i%6].hist(gene_10_fisher.values(), bins=bin_nums)\n",
    "    axs[i//6,i%6].set_title(classes[i],fontsize=8)\n",
    "    fig.suptitle(\"Fisher score distribution of the top 10% attention weighted genes for each cancer type\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Produces a plot of the genes of a cancer type that are in the LOWER x percent of attention weights and have a gene count greater than y.\n",
    "print(\"The top 10 of the botttom 90% genes based on attention weight for each cancer type:\")\n",
    "table_data = []\n",
    "for i in range(len(classes)):\n",
    "    top_gene_labels, cancer_label, gene_counts = get_top_gene_labels(i, 0.0, 0.9, True, False)\n",
    "    top_gene_labels =  [str(list(top_gene_labels)[i]) + \" \\n(\" +  str(list(gene_counts)[i]) + \")\" for i in range(len(list(top_gene_labels)))]\n",
    "    row = [cancer_label]\n",
    "    row.extend(top_gene_labels)\n",
    "    while len(row) < 11:\n",
    "        row.append(\"\")\n",
    "    table_data.append(np.array(row))\n",
    "    #print(row)\n",
    "    \n",
    "%matplotlib\n",
    "plt.figure(figsize=(16,8))\n",
    "widths = [1/(len(table_data[0])+5) for i in range(len(table_data[0]))]\n",
    "widths[0] = 5/(len(table_data[0])+5)\n",
    "table = plt.table(cellText=table_data,loc=\"center\", colWidths=widths)\n",
    "table.auto_set_font_size(False)\n",
    "table.set_fontsize(8)\n",
    "table.scale(1,2)\n",
    "\n",
    "plt.axis(\"off\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Produces a Matrix of histogram of the frequency of genes in the top 10% attention weighted genes for each cancer type\n",
    "%matplotlib\n",
    "fig,axs = plt.subplots(5,6)\n",
    "for i in range(27):\n",
    "    w, f = get_weights_and_genes(i, 0.9, 1.0, False)\n",
    "    \n",
    "    #for plotting:\n",
    "    filtered_instance_attention_genes = np.concatenate(np.expand_dims(f,axis=0)[0])\n",
    "\n",
    "    bin_nums = 100\n",
    "    #plotting\n",
    "    axs[i//6,i%6].hist(filtered_instance_attention_genes, bins=bin_nums)\n",
    "    axs[i//6,i%6].set_title(classes[i],fontsize=8)\n",
    "    fig.suptitle(\"Number of genes with attention weights in top 10%\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Produces a matrix of histograms representing the frequency of attention weights for hotspot instances and non-hotspot instances for all cancer types.\n",
    "%matplotlib\n",
    "fig,ax = plt.subplots(5,6)\n",
    "for i in range(27):\n",
    "    t_, h_, nh_ = get_position_weight_hotspots(i, five_p) \n",
    "    c1 = 0;c2 = 0;c3 = 0\n",
    "    for n in t_:\n",
    "        c1 += len(n)\n",
    "    for n in h_:\n",
    "        c2 += len(n)\n",
    "    for n in nh_:\n",
    "        c3 += len(n)\n",
    "\n",
    "    perc_hotspot = str(np.round_(100*c2/c1, 3))\n",
    "\n",
    "    bin_nums = 200\n",
    "    hotspot_att_weights = []\n",
    "    for weights in h_:\n",
    "        hotspot_att_weights.extend(weights)\n",
    "    non_hotspot_att_weights = []\n",
    "    for weights in nh_:\n",
    "        non_hotspot_att_weights.extend(weights)\n",
    "\n",
    "    #fig, ax = plt.subplots(figsize=(12, 6))\n",
    "    ax[i//6,i%6].hist(non_hotspot_att_weights, bins=bin_nums, label=\"non hotspots attention weights\")\n",
    "    ax[i//6,i%6].hist(hotspot_att_weights, bins=bin_nums, label=\"hotspots attention weights\")\n",
    "    ax[i//6,i%6].set_xlabel('attention weights')\n",
    "    ax[i//6,i%6].set_ylabel('Count')\n",
    "    ax[i//6,i%6].set_title(classes[i] + \"; \" + perc_hotspot + \"%\", fontsize=8)\n",
    "#fig.legend()\n",
    "fig.suptitle(\"Gene HotSpots Attention Weight Distribution (orange is hotspots)\")\n",
    "    #plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Produces a matrix of histograms representing the frequency of attention weights for indel instances and non-indel instances for all cancer types.\n",
    "%matplotlib\n",
    "fig,ax = plt.subplots(5,6)\n",
    "for i in range(27):\n",
    "    t_, h_, nh_ = get_position_weight_hotspots(i,genes)\n",
    "    c1 = 0;c2 = 0;c3 = 0\n",
    "    for n in t_:\n",
    "        c1 += len(n)\n",
    "    for n in h_:\n",
    "        c2 += len(n)\n",
    "    for n in nh_:\n",
    "        c3 += len(n)\n",
    "\n",
    "    perc_hotspot = str(np.round_(100*c2/c1, 3))\n",
    "\n",
    "    bin_nums = 200\n",
    "    hotspot_att_weights = []\n",
    "    for weights in h_:\n",
    "        hotspot_att_weights.extend(weights)\n",
    "    non_hotspot_att_weights = []\n",
    "    for weights in nh_:\n",
    "        non_hotspot_att_weights.extend(weights)\n",
    "\n",
    "    #fig, ax = plt.subplots(figsize=(12, 6))\n",
    "    ax[i//6,i%6].hist(non_hotspot_att_weights, bins=bin_nums, label=\"non indels attention weights\")\n",
    "    ax[i//6,i%6].hist(hotspot_att_weights, bins=bin_nums, label=\"indels attention weights\")\n",
    "    ax[i//6,i%6].set_xlabel('attention weights')\n",
    "    ax[i//6,i%6].set_ylabel('Count')\n",
    "    ax[i//6,i%6].set_title(classes[i] + \"; \" + perc_hotspot + \"%\", fontsize=8)\n",
    "#fig.legend()\n",
    "fig.suptitle(\"Gene InDel Attention Weight Distribution (orange is hotspots)\")\n",
    "    #plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
